### PR TITLE
Add recursive findfirst method for tuples.

### PR DESCRIPTION
**Pull request by chriselrod**
https://discourse.julialang.org/t/why-does-findfirst-t-on-a-tuple-of-typed-only-constant-fold-for-the-first/68893

With this PR:
```julia
julia> position(::T) where T = findfirst(==(T), (Int, Float64, Char))
position (generic function with 1 method)

julia> @code_typed position(10)
CodeInfo(
1 ─     return 1
) => Int64

julia> @code_typed position(10.0)
CodeInfo(
1 ─     return 2
) => Int64

julia> @code_typed position('j')
CodeInfo(
1 ─     return 3
) => Int64

julia> @code_typed position(10f0)
CodeInfo(
1 ─     return nothing
) => Nothing
```

### CODE REVIEW

**Comment by KristofferC** [comment]
Note https://github.com/JuliaLang/julia/pull/42263

File: base/tuple.jl [hunk 1]
```diff
@@ -354,6 +354,16 @@ _totuple(::Type{Tuple}, itr::NamedTuple) = (itr...,)
...
 
+## findfirst ##
+_findfirst(f, i::Int, ::Tuple{}) = nothing
+_findfirst(f, i::Int, x::Tuple) = (@inline; f(first(x)) ? i : _findfirst(f, i+1, tail(x)))
+findfirst(f::Function, x::Tuple) = _findfirst(f, 1, x)
+function findfirst(f::Function, x::Any32)
+  for i in 1:length(x)
+    f(x[i]) && return i
+  end
+end
```
Comment (aviatesk) [inline base/tuple.jl:365]: Let's use 4-spaces indents for `Base`:
```suggestion
function findfirst(f::Function, x::Any32)
    for i in 1:length(x)
        f(x[i]) && return i
    end
end
```

**Review by aviatesk — COMMENTED** [review]
You can use e.g.:
```julia
@test Base.return_types() do
    findfirst(==(2), (1,2,3))
end == Any[Int]
```
as a test case that is supposed to be improved by this PR.

Also, how about adding `findlast` also ? It might be enough to add:
```julia
function findlast(f::Function, x::Tuple)
    r = findfirst(f, reverse(x))
    return isnothing(r) ? r : length(x) - r + 1
end
```

File: base/tuple.jl (hunk 1 above)
Comment (chriselrod) [reply base/tuple.jl]: Done, switched to 4 spaces and renamed to `findfirst_loop` following #42263.

File: base/tuple.jl [hunk 2]
```diff
@@ -354,6 +354,21 @@ _totuple(::Type{Tuple}, itr::NamedTuple) = (itr...,)
 
 end
 
+## findfirst ##
```
Comment (aviatesk) [inline base/tuple.jl:357]:
```suggestion
## find ##

```

File: base/tuple.jl [hunk 3]
```diff
@@ -354,6 +354,21 @@ _totuple(::Type{Tuple}, itr::NamedTuple) = (itr...,)
...
+_findfirst_rec(f, i::Int, ::Tuple{}) = nothing
+_findfirst_rec(f, i::Int, t::Tuple) = (@inline; f(first(t)) ? i : _findfirst(f, i+1, tail(t)))
```
Comment (aviatesk) [inline base/tuple.jl:359]:
```suggestion
_findfirst_rec(f, i::Int, t::Tuple) = (@inline; f(first(t)) ? i : _findfirst_rec(f, i+1, tail(t)))
```

File: base/tuple.jl [hunk 4]
```diff
@@ -354,6 +354,21 @@ _totuple(::Type{Tuple}, itr::NamedTuple) = (itr...,)
...
+function _findfirst_loop(f::Function, t)
+    for i in 1:length(t)
+        f(t[i]) && return i
+    end
```
Comment (aviatesk) [inline base/tuple.jl:363]:
```suggestion
    end
    return nothing
```

File: test/tuple.jl [hunk 5]
```diff
@@ -524,6 +524,10 @@ end
         @test findnext(isequal(1), (1, 1), UInt(2)) isa Int
         @test findprev(isequal(1), (1, 1), UInt(1)) isa Int
     end
+
+    @test Base.return_types() do
+        findfirst(==(2), (1.0,2,3f0))
+    end == Any[Int]
```
Comment (aviatesk) [inline test/tuple.jl:530]:
```suggestion
    # recursive implementation should allow constant-folding for small tuples
    @test Base.return_types() do
        findfirst(==(2), (1.0,2,3f0))
    end == Any[Int]
    @test Base.return_types() do
        findfirst(==(0), (1.0,2,3f0))
    end == Any[Nothing]
    @test Base.return_types() do
        findlast(==(2), (1.0,2,3f0))
    end == Any[Int]
    @test Base.return_types() do
        findlast(==(0), (1.0,2,3f0))
    end == Any[Nothing]
```

**Comment by aviatesk** [comment]
Let's see a benchmark result: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/42423#issuecomment-930521771):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(setenv(`git push`; dir="/run/media/system/data/nanosoldier/workdir/NanosoldierReports"), ProcessExited(1)) [1]

```
Unfortunately, the logs could not be uploaded.
cc @christopher-dG

**Comment by aviatesk** [comment]
@vtjnash could you please upload the result ?

**Comment by vtjnash** [comment]
https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/2976c42_vs_7f26a7f/report.md

yep, looks unchanged to me

**Comment by aviatesk** [comment]
Thanks. And yeah, the result looks reasonable and I think this is good to go.

**Review by aviatesk — APPROVED** [review]

File: base/tuple.jl [hunk 6]
```diff
@@ -354,6 +354,23 @@ _totuple(::Type{Tuple}, itr::NamedTuple) = (itr...,)
...
+
+_findfirst_rec(f, i::Int, ::Tuple{}) = nothing
+_findfirst_rec(f, i::Int, t::Tuple) = (@inline; f(first(t)) ? i : _findfirst_rec(f, i+1, tail(t)))
+function _findfirst_loop(f::Function, t)
+    for i in 1:length(t)
+        f(t[i]) && return i
+    end
+    return nothing
+end
+findfirst(f::Function, t::Tuple) = length(t) < 32 ? _findfirst_rec(f, 1, t) : _findfirst_loop(f, t)
```
Comment (KristofferC) [inline base/tuple.jl:367]: I guess this should be `<=` if we want to include the 32 case?

Comment (chriselrod) [reply base/tuple.jl:367]: I went with `<` because Jeff did here: https://github.com/JuliaLang/julia/pull/42263/files

**Comment by aviatesk** [comment]
All CI failures are known issues.

**Pull request merged**